src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1863,11 +1875,11 @@ protected virtual void MarkMulticastDelegate (TypeDefinition type)
 			MarkMethodCollection (type.Methods, new DependencyInfo (DependencyKind.MethodForSpecialType, type));
 		}
 
-		protected (TypeReference, DependencyInfo) GetOriginalType (TypeReference type, DependencyInfo reason)
+		protected (TypeReference, DependencyInfo) GetOriginalType (TypeReference type, DependencyInfo reason, IMemberDefinition sourceLocationMember)
 		{
 			while (type is TypeSpecification specification) {
 				if (type is GenericInstanceType git) {
-					MarkGenericArguments (git);
+					MarkGenericArguments (git, sourceLocationMember ?? (reason.Source as IMemberDefinition) ?? (reason.Source as MemberReference)?.Resolve ());
```

Comment:
I'd prefer if `reason.Source` was not used in this context as I have no confidence in the value and we have no tests for that

Reply:
In that case we would have to pass the new parameter on every call to MarkType and I would have to add a similar parameter to MarkMethod and also pass it everywhere.

I actually did that at first, but it turned out that in almost all cases we passed the same value to that new parameter as we do to reason.Source.

So in a way the tests for this new functionality would also validate the reason.Source.

Reply:
I propagated the source member explicitly - this spreads everywhere, so the change is large (almost all of mark step), but mostly it's just mechanical.